packages/coding-standard/src/Rules/NoFactoryInConstructorRule.php

```diff
@@ -22,45 +25,52 @@ final class NoFactoryInConstructorRule extends AbstractManyNodeTypeRule
     public const ERROR_MESSAGE = 'Do not use factory in constructor';
 
     /**
-     * @var NodeFinder
+     * @var string[]
      */
-    private $nodeFinder;
-
-    public function __construct(NodeFinder $nodeFinder)
-    {
-        $this->nodeFinder = $nodeFinder;
-    }
+    private const ALLOWED_TYPES = [ParameterProvider::class, ParameterBagInterface::class];
 
     /**
      * @return string[]
      */
     public function getNodeTypes(): array
     {
-        return [ClassMethod::class];
+        return [MethodCall::class];
     }
 
     /**
-     * @param ClassMethod $node
+     * @param MethodCall $node
      * @return string[]
      */
     public function process(Node $node, Scope $scope): array
     {
-        if ((string) $node->name !== '__construct') {
+        $reflectionFunction = $scope->getFunction();
+        if (! $reflectionFunction instanceof MethodReflection) {
+            return [];
+        }
+
+        if ($reflectionFunction->getName() !== '__construct') {
+            return [];
+        }
+
+        if (! $node->var instanceof Variable) {
             return [];
         }
 
-        $methodCalls = $this->nodeFinder->findInstanceOf((array) $node->getStmts(), MethodCall::class);
-        /** @var MethodCall $methodCall */
-        foreach ($methodCalls as $methodCall) {
-            /** @var Variable $variable */
-            $variable = $methodCall->var;
-            $name = $variable->name;
```

Comment:
@samsonasik Btw, this is dangerous to assume, as it breaks code later.

```php
/** @var Variable $variable */
$variable = $methodCall->var; // this can be anything, MethodCall, PropertyFetch etc.
```

It's better to always check with instanceof:

```php
$methodCaller = $methodCall->var;
if (! $methodCaller instanceof Variable) {
    continue;
}
```